app/code/Magento/CatalogInventory/Model/StockStateProvider.php

```diff
@@ -106,6 +106,10 @@ public function checkQuoteItemQty(StockItemInterface $stockItem, $qty, $summaryQ
     {
         $result = $this->objectFactory->create();
         $result->setHasError(false);
+        
+        $product = $this->productFactory->create();
+        $product->load($stockItem->getProductId());
```

Comment:
If you need to load Product you have to rely on ProductRepostioryInterface::get method.
loading product using the **load** method of Product Model you are using deprecated method of Abstract Model https://github.com/magento/magento2/blob/develop/lib/internal/Magento/Framework/Model/AbstractModel.php#L526